### PR TITLE
Public schema support

### LINKED ISSUE #407: Not able to use schemas added by another user
(Feedback from FSA phase 3 testing)

### PR DESCRIPTION
**Pull request by Floppy**
This PR fixes #407.

Changes proposed in this pull request:
- Added a `restricted` flag for `DatasetFileSchema`. Named to match the flag in `Dataset`.
- Access control option everywhere in the UI that schemas are created or edited.
  - New dataset page
  - Edit dataset page
  - New schema page
  - Edit schema page
  - Infer schema page
- Public schemas are listed in a separate list on the schema index page
- Public schemas are available in all schema selection dropdowns

### CODE REVIEW

File: app/controllers/dataset_file_schemas_controller.rb [hunk 1]
```diff
@@ -4,6 +4,7 @@ class DatasetFileSchemasController < ApplicationController
 
   def index
     @dataset_file_schemas = DatasetFileSchema.where(user: current_user).order(created_at: :desc)
+    @public_schemas = DatasetFileSchema.where(restricted: false).order(created_at: :desc)
```
Comment (quadrophobiac) [inline app/controllers/dataset_file_schemas_controller.rb]: does this mean schemas already in the database will have the restricted flag disabled by default?

File: app/services/dataset_file_schema_service.rb [hunk 2]
```diff
@@ -1,19 +1,20 @@
 class DatasetFileSchemaService
 
-  def initialize(schema_name, description, url_in_s3, user, owner_username = user.name)
+  def initialize(schema_name, description, url_in_s3, user, owner_username = user.name, restricted = true)
     @schema_name = schema_name
     @description = description
     @url_in_s3 = url_in_s3
     @user = user
     @owner_username = owner_username
+    @restricted = restricted
```
Comment (quadrophobiac) [inline app/services/dataset_file_schema_service.rb]: does this mean schemas already in the database will have the restricted flag disabled by default?

Comment (Floppy) [inline app/services/dataset_file_schema_service.rb]: Yes, both here and in the database migration itself, restricted is by default `true` to preserve existing behaviour.

**Comment by Floppy** [comment]
I've got one more bit to go in this, not quite done yet.

**Comment by quadrophobiac** [comment]
apologies just saw the email notification, will return when you remove [WIP]

**Comment by Floppy** [comment]
No problem, I requested the review a bit early by accident. :)

**Comment by Floppy** [comment]
This is OK to go now. There was one bit of mocking I couldn't work out which currently has two pending tests instead. But the functionality does work in manual test, and I think I've tested at the controller level if not at the feature level.

**Comment by Floppy** [comment]
Note that there is more of this to come with private schema sharing (#558, to share schemas with specific users), but getting this merged in will be a good base to do that from, and there's a few other bits to do first.

**Comment by Floppy** [comment]
Sorry this is large - the entire diff might be a bit overwhelming. The commits, however, are quite orderly. It might be easier to review those in order, they tell the story more accurately I reckon.

**Comment by quadrophobiac** [comment]
So the code looks good but I have a question: given that 
> Public schemas are available in **all** schema selection dropdowns
would making schema description mandatory for any schema listed as public be helpful? Then that description could maybe made available when it's selected from the dropdown menu (or on a tooltip hover - I'm literally thinking aloud here) 
Basically I'm just picturing a few steps down the line where there are potentially a lot of publicly available schemas and addressing how an octopub user will know how to sort through and select from a long dropdown menu

**Comment by quadrophobiac** [comment]
@Floppy this is reviewed and fine (I couldn't find a way to indicate that in this conversation) so if you want to merge it as is go ahead, my above query can be a separate issue ticket if you think it merits it

**Comment by Floppy** [comment]
Yep, that's a good idea @quadrophobiac - add it as a future enhancement.

**Pull request merged**